pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,281 @@
+            if k.startswith('pipeline-'):
+                i += 1
+                configs = self.cfg.getliteral(cfgsect, k)
+                configs = [configs] if isinstance(configs, dict) else configs
+                pname = k[9:]
+
+                pipel_cfg[pname] = configs
+                params = ConduitNode()
+                for (j, filter) in enumerate(configs):
+                    for kf in filter:
+                        if kf == 'type':
+                            pl[f'pl{i}/f{j}/{kf}'] = filter[kf]
+                        else:
+                            params[f'{kf}'] = filter[kf]
+
+                        if kf == 'field':
+                            try:
+                                _ = self._expr[kf]
+                            except KeyError:
+                                raise KeyError(f"Missing field '{kf}' used in pipeline")
+                pl[f'pl{i}/f{j}/params'] = params
+
+        self.pipelines = pl
+        return pipel_cfg
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        # Iterate over each element type in the simulation
+        for i, (soln, (ename, eles)) in enumerate(zip(intg.soln, 
+                                                  self._ele_map.items())):
+            
+            dom_str = f'domain_{self.dom_0 + i}/'
+            # Subset and transpose the solution
+            soln = soln[...,:].swapaxes(0, 1)
+
+            mesh_op, soln_op = self.ops[ename]
+            # Interpolate the solution to the subdivided points
+            soln = soln_op @ soln
+
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln, self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+            # subs |= dict(zip('xyz', plocs)) # TODO: add the ability for xyz in functions
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                # Compute the gradients
+                grad_soln = np.rollaxis(intg.grad_soln[i], 2)[..., :]
+
+                # Interpolate the gradients to the quadrature points
+                grad_soln = soln_op @ grad_soln
+
+                # Transform from conservative to primitive gradients
+                pgrads = self.elementscls.grad_con_to_pri(soln, grad_soln,
+                                                          self.cfg)
+
+                # Add them to the substitutions dictionary
+                for pname, idx in self._gradpinfo:
+                    for dim, grad in zip('xyz', pgrads[idx]):
+                        subs[f'grad_{pname}_{dim}'] = grad
+
+            for (j, field) in enumerate(self._expr):
+                # Evaluate the expression at each point
+                fun = npeval(self._expr[field], subs)
+                (nupts, neles) = np.shape(fun)
```

Comment:
Not sure why this is needed?

Reply:
Which bit? I found that I needed to reshape the input to get this to work.

Reply:
Reshaping to 1D should never be needed.  More likely something like `ascontiguousarray` is needed.